### ISSUE TITLE
ReferenceManyField source=<Array>

### ISSUE THREAD

**Issue by suricactus** [marmelab/admin-on-rest#429]
I'm building a new REST API with many-to-many relationship. In this case it would be awesome if ReferenceManyField could work with list of ids in base object (using GET_MANY instead of GET_MANY_REFERENCE). Here is an example:

Get a list of posts with a list of their categories:
```
/posts
[
{
  "title": "My Awesome Post",
  "categories": [ 2, 3 ]
},
{
  "title": "My Second Post",
  "categories": [ 3, 4 ]
},
]
```
Then if we are in ListView we call GET_MANY for the distinct of all category's ids:
```
/categories?filter={ids:[2,3,4]}
[
{ id: 2, name: "Second Category"},
{ id: 3, name: "Third Category"},
{ id: 4, name: "Fourth Category"}
]
```
Finally we render them in our list view as simple as this:
```
<ReferenceManyField reference="categories" source="props.record.category">
  <SingleFieldList>
    <ChipField source="name" />
  </SingleFieldList>
</ReferenceManyField>
```

**Comment by fzaninotto** [comment]
That should be another component, probably called `<ReferenceArrayField>`. Using `<ReferenceManyField>` as an example, I think it's quite easy to write, and I advise you to give it a try. If you're satisfied with your code, send us a PR!

**Comment by suricactus** [comment]
I am new to React world, but I'll take this opportunity to learn it faster and better by trying to implement such component.

**Comment by leesei** [comment]
I've make this based on `<ReferenceArrayField>`.

```
// ReferenceArrayField.js
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import LinearProgress from 'material-ui/LinearProgress';
import { crudGetMany as crudGetManyAction } from 'admin-on-rest';

/**
 * Render records specified by a array of ids in current one.
 *
 * You must define the fields to be passed to the iterator component as children.
 *
 * @example Display all the products of the current order as datagrid
 * <ReferenceArrayField reference="products" target="product_ids">
 *     <Datagrid>
 *         <TextField source="id" />
 *         <TextField source="title" />
 *         <DateField source="price" />
 *         <EditButton />
 *     </Datagrid>
 * </ReferenceArrayField>
 *
 * @example Display all the categories of the current product as a list of chips
 * <ReferenceArrayField label="Category" reference="category" source="category_ids">
 *     <SingleFieldList>
 *         <ChipField source="name" />
 *     </SingleFieldList>
 * </ReferenceArrayField>
 *
 */
class ReferenceArrayField extends Component {
    constructor(props) {
        super(props);
    }

    componentDidMount() {
        const { crudGetMany, reference, ids } = this.props;
        crudGetMany(reference, ids);
    }

    render() {
        const { data, ids, children, resource, reference, basePath } = this.props;
        if (React.Children.count(children) !== 1) {
            throw new Error('<ReferenceArrayField> only accepts a single child (like <Datagrid>)');
        }
        if (typeof ids === 'undefined') {
            return <LinearProgress style={{ marginTop: '1em' }} />;
        }

        const referenceBasePath = basePath.replace(resource, reference); // FIXME obviously very weak
        return React.cloneElement(children, {
            resource: reference,
            ids,
            data,
            basePath: referenceBasePath
        });
    }
}
ReferenceArrayField.propTypes = {
    addLabel: PropTypes.bool,
    basePath: PropTypes.string.isRequired,
    children: PropTypes.element.isRequired,
    crudGetMany: PropTypes.func.isRequired,
    ids: PropTypes.array,
    label: PropTypes.string,
    record: PropTypes.object,
    reference: PropTypes.string.isRequired,
    data: PropTypes.object,
    resource: PropTypes.string.isRequired,
    source: PropTypes.string.isRequired,
};

const mapStateToProps = (state, props) => {
    const { record, source, reference } = props;
    const ids = record[source];
    return {
        ids: ids,
        data: ids
            .map(id => state.admin[reference].data[id])
            .filter(r => typeof r !== 'undefined')
            .reduce((prev, next) => {
                prev[next.id] = next;
                return prev;
            }, {}),
    };
};

const ConnectedReferenceArrayField = connect(mapStateToProps, {
    crudGetMany: crudGetManyAction,
})(ReferenceArrayField);

ConnectedReferenceArrayField.defaultProps = {
    addLabel: true,
    source: '',
};

export default ConnectedReferenceArrayField;
```

**Comment by fzaninotto** [comment]
@leesei That looks great! Why don't you open a PR?

**Comment by velociwabbit** [comment]
I am struggling with a similar problem and think that this implementation might work for various permutations of charting within a datagrid.

I have records where each id has a "now" record and history records associated with it (these are a set of sensors that generate values every few seconds.)

What I would like to do is create a datagrid with filter and lensing (i am not sure if that is the correct term  but the concept would be to drill down on any sensors current value (which is displayed as a dial))  and view its history over the last hour day week etc.

The other option would be to enable the user to select a set of sensors and view that data both aggregated as a summary and individually within a given time frame.

I keep circling this problem and not really making progress.    Do you think this component could be used to implement such a solution? 

And if so how would you wire it up through crudRoute, saga, redux etc?

The reason why I am asking here is that this is a new component and my question might be useful to elaborate on how to use it for others in the future.

**Comment by leesei** [comment]
@fzaninotto will do

@velociwabbit are you familiar with the concept of [Presentational and Container Components](http://redux.js.org/docs/basics/UsageWithReact.html#presentational-and-container-components)?
`<RefereceArrayField>` here is a Container Component, it uses actions from aor and the cache `state.admin[reference]` to fetch data from server.
You can actually create a Container Component that fetches all you data you need and then render them in the your children (your React Presentational Components). In this sense aor only provides the action and store schema (which are helpful), the rest is Redux-fu and React-fu.

**Comment by velociwabbit** [comment]
Thanks for the tip.

 I was struggling last night trying to shoe horn my grid-based class into  SensorList as an aor List component and while I was able to render the page, filtering it and changing the grid displays turned out to be a bit of a challenge :). 

As an aside i am a bit underwhelmed with the complexity/ performance value of both saga and redux forms. I dont know if anyone using functional style programming remembers or even tried to program using Microsofts c/ c++ MFC libraries but a similar message based processing pattern is used for them as well.    

The problem with message passing is envisioning the message order and paths and figuring out where things go south. And while id does abstract the software into a set of parameters it does not reduce the overall amount of effort and loc required to develop/enhance/maintain the system.

If there are other alternatives I will gladly be the pioneer to see what  little bit of value I can add to this project.

**Comment by leesei** [comment]
@velociwabbit Same thought here.
In studying a large Redux project like aor I found that you don't have to know whose is responding to the action as long as you know where to look for the data. This is a blessing when everything works fine.
I too was hit by cryptic saga error when error occurs. But I confess I've yet to dive deep into saga and its still easy to trace the code path given you know which action caused the error.

Anyway I believe you don't *have* to use saga and redux-form, you can use plain redux reducer and custom action to update your slice of state.

**Comment by velociwabbit** [comment]
Because the project that I am working on is a real time and potentially mission critical app I have had to really drill down into every module and library I am using and make sure that the number of unknowns caused by extra libraries and code bloat is minimized. 

It is an exercise in humility, confusion, frustration, relief and then eventually a bit of  self righteousness when I realize that I can eliminate several layers of react components without losing functionality.  

This happened today when i followed up on fzaninotto's  really good suggestion about   component vs container-based React classes.  I noticed in the aor master example gridlist and gridtile are not used. I now know why. (What can be done in 2 lines of html should not require 200 lines of React!) 

I also learned that the hardest part to both understand and "reduce"  are the redux/forms/saga stuff.  This does not bode well for them long term.  If something is hard to master (for example COM because it was so finicky) eventually, (and no matter who is promoting it)  it will be sidestepped by the market.

This is also compounded by modular design which makes looking at the "big picture' and stepping through the "state table" all but impossible.

Maybe what is needed is a ladder logic equivalent for web software ( I am joking of course and trying to make a not so subtle point!)

There has to be a better way, I just don't know what it is yet.
